Add validation case 2e in TMAP8 (#236)
idaholab/TMAP8: doc/content/verification_and_validation/val-2e.md

## Model Description

In first three cases, TMAP8 simulates a one-dimensional domain to represent the deuterium diffusion and gas flowing in enclosure.
Note that this case can easily be extended to a two- or three-dimensional case.

This model considers the diffusion of deuterium on both sides of Pd membrane.
First, the diffusion of deuterium in membrane is described as:

\begin{equation} \label{eq:diffusion}
\frac{d C_i}{d t} = \nabla D_i \nabla C_i,
\end{equation}

where $C_i$ is the concentration of hydrogen isotope $i$ in membrane,
$i$ represents hydrogen isotope H or D,
$t$ is the time,
$D_i$ is the diffusivity of hydrogen isotope $i$.

The flow of gas into enclosures 2 and 3 can be given by

\begin{equation} \label{eq:dt_P2}
\frac{P_{Ij}}{dt} = \frac{Q (P_{I,j-1}-P_{Ij})}{V_j},
\end{equation}

where $Q$ is the volumetric flow rate of always 0.1 m$^3$/s,
$V_j$ is the volume of current enclosure $j$, $P_{I,j-1}$ and $P_{I,j}$ are the pressure of gas molecules $I$ in the previous enclosure and current enclosure, gas molecules $I$ represents H$_2$, or D$_2$, or HD.

> The hydrogen isotope recombines into gas on both sides of the membrane.
> By assuming that the recombination process is at steady state, the concentrations on both sides are describe by
> 
> \begin{equation} \label{eq:seivert_concentration}
> C_i = K_s P_{Ij}^{n},
> \end{equation}
> 
> where $K_s$ is the solubility of hydrogen isotope in membrane, and $n$ is the exponent for the relation of pressure and concentration.
> 
> The co-permeation simulation with lawdep uses the same sievert boundary condition,
> whereas, the simulation with ratedep uses the recombination boundary condition, which is described as
> 
> \begin{equation} \label{eq:recombination}
> J_{ij} = 2 A (K_r C_i^2 - K_d P_{Ij}),
> \end{equation}
> 
> where $J_{ij}$ is the recombination flux of hydrogen isotope $i$ out of the sample sides in enclosure $j$,
> $A$ is the area on the upstream or downstream side,
> and $K_r$ and $K_d$ are the recombination and dissociation coefficients, respectively.
> The coefficient of 2 accounts for the fact that 2 atoms form one isotope molecule.
> 
> The two co-permeation simulations also include the chemical reactions on upstream and downstream enclosure.
> The reaction rates for H$_2$, D$_2$, and HD are described by
> 
> \begin{equation} \label{eq:chemical_reaction_lawdep}
> \frac{- 2 d P_{H_2}}{dt} = \frac{- 2 d P_{D_2}}{dt} = \frac{d P_{HD}}{dt} = 2 P_{H_2}^{0.5} P_{D_2}^{0.5}
> \end{equation}
> 
> for lawdep boundary condition and
> 
> \begin{equation} \label{eq:chemical_reaction_ratedep}
> \frac{- 2 d P_{H_2}}{dt} = \frac{- 2 d P_{D_2}}{dt} = \frac{d P_{HD}}{dt} = \frac{k_b T A}{V} (K_r C_H C_D - K_d P_{HD}),
> \end{equation}
> 
> for ratedep boundary condition,
> where $k_b$ is the Boltzmann’s constant,
> $A$ is the surface area of the membrane,
> $T$ are the temperature.

Reviewer: ```suggestion
The hydrogen isotopes recombine into gas on both sides of the membrane, and there are several ways to model this process, two of which are used in this validation case: (1) Assuming steady state at the surface, also called lawdep boundary conditions in TMAP4 and TMAP7 [!citep](ambrosek2008verification), and (2) capturing the kinetics of dissolution and recombination, also called ratedep conditions in TMAP4 and TMAP7 [!citep](ambrosek2008verification).

By assuming that the kinetics of the dissolution and recombination processes are faster than diffusion, surface reactions can be assumed to be at steady state.
The concentrations on both sides can then be described as

\begin{equation} \label{eq:seivert_concentration}
C_i = K_s P_{Ij}^{n},
\end{equation}

where $K_s$ is the solubility of hydrogen isotope in the membrane, and $n$ is the exponent for the relation of pressure and concentration.

However, for the kinetics of surface reactions to be captured, the boundary conditions is set as

\begin{equation} \label{eq:recombination}
J_{ij} = 2 A (K_r C_i^2 - K_d P_{Ij}),
\end{equation}

where $J_{ij}$ is the recombination flux of hydrogen isotope $i$ out of the sample sides in enclosure $j$,
$A$ is the area on either side of the membrane,
and $K_r$ and $K_d$ are the recombination and dissociation coefficients, respectively.
The coefficient of 2 accounts for the fact that 2 atoms (D and H) form one isotope molecule (HD).

The two co-permeation simulations, i.e., val-2ed and val-2ec, also include the chemical reactions in the upstream and downstream enclosure.
The reaction rates for H$_2$, D$_2$, and HD are described by

\begin{equation} \label{eq:chemical_reaction_lawdep}
\frac{- 2 d P_{H_2}}{dt} = \frac{- 2 d P_{D_2}}{dt} = \frac{d P_{HD}}{dt} = 2 P_{H_2}^{0.5} P_{D_2}^{0.5}
\end{equation}

for lawdep boundary condition and

\begin{equation} \label{eq:chemical_reaction_ratedep}
\frac{- 2 d P_{H_2}}{dt} = \frac{- 2 d P_{D_2}}{dt} = \frac{d P_{HD}}{dt} = \frac{k_b T A}{V} (K_r C_H C_D - K_d P_{HD}),
\end{equation}

for ratedep boundary condition,
where $k_b$ is the Boltzmann’s constant,
$A$ is the surface area of the membrane,
$T$ are the temperature.
```

| $P_{D_2,3}$ | initial D$_2$ pressure in enclosure 3 | 1 $\times 10^{-20}$                                     | Pa                    |
| $P_{H_2,3}$ | initial H$_2$ pressure in enclosure 3 | 1 $\times 10^{-20}$                                     | Pa                    |
| $P_{HD,3}$  | initial HD pressure in enclosure 3   | 1 $\times 10^{-20}$                                      | Pa                    |
| $P_{D_2,4}$ | D$_2$ pressure in enclosure 4        | 1 $\times 10^{-10}$                                      | Pa                    |
| $P_{H_2,4}$ | H$_2$ pressure in enclosure 4        | 1 $\times 10^{-10}$                                      | Pa                    |
| $P_{HD,4}$  | HD pressure in enclosure 4           | 1 $\times 10^{-10}$                                      | Pa                    |
| $P_{H_2,5}$ | H$_2$ pressure in enclosure 4        | 6.3 $\times 10^{-2}$                                     | Pa                    |
| $P_{HD,5}$  | HD pressure in enclosure 4           | 1 $\times 10^{-10}$                                      | Pa                    |

> !alert note title=Updates of times and initial pressures in val-2ee to be consistent with val-2ed.
> In [!cite](ambrosek2008verification), the times and initial pressures in val-2ee are different from these values in val-2ed. However, we use the same values from val-2ed in val-2ee for consistency. The trivial difference on times and initial pressures will not impact the final results.

Reviewer: ```suggestion
!alert note title=Updates of times and initial pressures in val-2ee to be consistent with val-2ed.
In [!cite](ambrosek2008verification), the times and initial pressures in val-2ee are different from the values used in val-2ed. However, we use the same values from val-2ed in val-2ee for consistency. The trivial difference in times and initial pressures do not impact the final results.
```

| $Q$   | flow rate in all enclosures         | 0.1                       | m$^3$/s                   | [!cite](ambrosek2008verification) |
| $L_1$ | thinckness of membrane in val-2ea     | 5$\times 10^{-5}$         | m                    | [!cite](kizu2001co) |
| $L_2$ | thinckness of membrane in val-2eb, val-2ec, val-2ed, val-2ee | 2.5$\times 10^{-5}$ | m          | [!cite](kizu2001co) |
| $D_H$ | hydrogen diffusivity in membrane         | 3.728$\times 10^{-4} \exp(-1315.8/T)$        | m$^2$/s              | [!cite](katz1960permeability) |
| $D_D$ | deuterium diffusivity in membrane         | 2.636$\times 10^{-4} \exp(-1315.8/T)$        | m$^2$/s              | [!cite](ambrosek2008verification) |
| $K_{s,1}$ | solubility of hydrogen isotope in val-2ea, val-2eb, val-2ec  | 1.511$\times 10^{23} \exp(-5918/T)$        | atom/m$^3$/Pa$^{0.9297}$  | [!cite](ambrosek2008verification) |
| $K_{s,2}$ | solubility of hydrogen isotope in val-2ed, val-2ee  | 9.355$\times 10^{22} \exp(-5918/T)$        | atom/m$^3$/Pa$^{0.9297}$  | [!cite](ambrosek2008verification) |
| $n$ | exponent for the relation of pressure and concentration | 0.9297         | -                     | [!cite](kizu2001co) |

> | $K_{r,D_2}$ | D$_2$ recombination coefficient | 2.502$\times 10^{-24} / \sqrt(4T) \exp(11836/T)$        | m$^4$/s              | [!cite](ambrosek2008verification) |
> | $K_{r,H_2}$ | H$_2$ recombination coefficient | 2.502$\times 10^{-24} / \sqrt(2T) \exp(11836/T)$        | m$^4$/s              | [!cite](ambrosek2008verification) |
> | $K_{r,HD}$ | HD recombination coefficient | 2.502$\times 10^{-24} / \sqrt(3T) \exp(11836/T)$        | m$^4$/s              | [!cite](ambrosek2008verification) |
> | $K_{d,D_2}$ | D$_2$ dissociation coefficient | 2.1897$\times 10^{22} / \sqrt(4T)$     | molecular/m$^2$/Pa              | [!cite](ambrosek2008verification) |
> | $K_{d,H_2}$ | H$_2$ dissociation coefficient | 2.1897$\times 10^{22} / \sqrt(2T)$     | molecular/m$^2$/Pa              | [!cite](ambrosek2008verification) |
> | $K_{d,HD}$ | HD dissociation coefficient | 2.1897$\times 10^{22} / \sqrt(3T)$     | molecular/m$^2$/Pa              | [!cite](ambrosek2008verification) |

Reviewer: ```suggestion
| $K_{r,D_2}$ | D$_2$ recombination coefficient | 2.502$\times 10^{-24} / \sqrt{4T} \exp(11836/T)$        | m$^4$/s              | [!cite](ambrosek2008verification) |
| $K_{r,H_2}$ | H$_2$ recombination coefficient | 2.502$\times 10^{-24} / \sqrt{2T} \exp(11836/T)$        | m$^4$/s              | [!cite](ambrosek2008verification) |
| $K_{r,HD}$ | HD recombination coefficient | 2.502$\times 10^{-24} / \sqrt{3T} \exp(11836/T)$        | m$^4$/s              | [!cite](ambrosek2008verification) |
| $K_{d,D_2}$ | D$_2$ dissociation coefficient | 2.1897$\times 10^{22} / \sqrt{4T}$     | molecular/m$^2$/Pa              | [!cite](ambrosek2008verification) |
| $K_{d,H_2}$ | H$_2$ dissociation coefficient | 2.1897$\times 10^{22} / \sqrt{2T}$     | molecular/m$^2$/Pa              | [!cite](ambrosek2008verification) |
| $K_{d,HD}$ | HD dissociation coefficient | 2.1897$\times 10^{22} / \sqrt{3T}$     | molecular/m$^2$/Pa              | [!cite](ambrosek2008verification) |
```

> ## Results

Reviewer: I have a couple key questions for this:
- Did you use the experimental results from the TMAP7 V&V manual directly or did you check the original paper directly? They sometimes have differences. 
- When plotting experimental data points (as opposed to a curve), then use a scatter plot for it. The model should be given as a continuous line, but the experimental data should only be given in local points. 
- (I  still need to check how you calculate it, nut I wanted to make this comment here to make the expectation clear and to remember to verify) When you determine the RMSPE, make sure to use only the experimental points that were collected and reported against their corresponding model prediction, rather than the points at each time step of the model against interpolated experimental data.

Author: 1. I extracted the data from TMAP7 in last commit, but you are right, it is better to extract data from original paper. So I checked and updated the experiment data with original paper in this commit
2. Done
3. I calculate RMSPE on experimental data points, and the simulation data are interpolated if they are different with experimental data points.

> [val-2e_comparison_diffusion] shows the comparison of the TMAP8 calculation and the experimental data from [!cite](kizu2001co), including simulations val-2ea, val-2eb, and val-2ec. There are reasonable agreements between the TMAP predictions and the experimental data with root mean square percentage errors of RMSPE = 22.17 %, 23.68 %, and 23.21 % for val-2ea, val-2eb, and val-2ec.

Reviewer: ```suggestion
[val-2e_comparison_diffusion] shows the comparison of the TMAP8 calculation and the experimental data from [!cite](kizu2001co), including simulations val-2ea, val-2eb, and val-2ec. There are reasonable agreements between the TMAP predictions and the experimental data with root mean square percentage errors (RMSPE) of RMSPE = 22.17 %, 23.68 %, and 23.21 % for val-2ea, val-2eb, and val-2ec.
```

> [val-2e_comparison_mixture_diffusion] shows the comparison of the D$_2$, H$_2$, and HD flux versus effective D$_2$ pressure on the upstream side of the membrane from TMAP8 calculation using lawdep condition and the experimental data. The effective D$_2$ pressure is calculated using $P_{D_2} + 0.5 P_{HD}$. There is not a good agreement except for H$_2$ at low pressure, and D$_2$ at high pressure. Which is similar to the results from TMAP7. The root mean square percentage errors are RMSPE = 66.07 %, 49.18 %, 119.67 %, 22.96 % for H$_2$, D$_2$, HD, and the summation of all gas.

Reviewer: ```suggestion
[val-2e_comparison_mixture_diffusion] shows the comparison of the D$_2$, H$_2$, and HD flux versus effective D$_2$ pressure on the upstream side of the membrane from TMAP8 calculation using lawdep condition and the experimental data. The effective D$_2$ pressure is calculated using $P_{D_2} + 0.5 P_{HD}$. There is not a good agreement except for H$_2$ at low pressure, and D$_2$ at high pressure. Which is similar to the results from TMAP7. In this scenario, RMSPE = 66.07 %, 49.18 %, 119.67 %, 22.96 % for H$_2$, D$_2$, HD, and the summation of all gas.
```

> [val-2e_comparison_mixture_diffusion_recombination] shows the comparison of the D$_2$, H$_2$, and HD flux versus effective D$_2$ pressure on the upstream side of the membrane from TMAP8 calculation using ratedep condition and the experimental data. There are reasonable agreement between the TMAP predictions and the experimental data with root mean square percentage errors of RMSPE = 14.14 %, 58.53  %, 85.22 %, 59.57  % for H$_2$, D$_2$, HD, and the summation of all gas.

Reviewer: ```suggestion
[val-2e_comparison_mixture_diffusion_recombination] shows the comparison of the D$_2$, H$_2$, and HD flux versus effective D$_2$ pressure on the upstream side of the membrane from TMAP8 calculation using ratedep condition and the experimental data. There are reasonable agreement between the TMAP predictions and the experimental data with RMSPE = 14.14 %, 58.53 %, 85.22 %, 59.57 % for H$_2$, D$_2$, HD, and the summation of all gas.
```

Note that the agreement could be improved by adjusting the model parameters. It is also possible to perform this optimization with [MOOSE's stochastic tools module](https://mooseframework.inl.gov/modules/stochastic_tools/index.html).

> !media comparison_val-2e.py

Reviewer: In these figures, it is easier for the reader to read which line is which when the legend, RMSPE, and lines on the graph are in the same order from top to bottom. You did it for RMSPE and the legend, but it does not correspond to the order of the data on the plot (note that you sometimes have to make an arbitrary decision when lines cross, but that's OK. Also, you should put the corresponding experimental and modeling legend entries together). Here, for example, the order of the legend and RMSPE list should be reversed to match the order of the lines.

Author: Done

>        image_name=val-2e_comparison_diffusion.png

Reviewer: Make sure to add ver-1ea,b,c in the legend (and do the same for the other figures).

Author: Done

       style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
       id=val-2e_comparison_diffusion
       caption= Comparison of the D$_2$ flux versus D$_2$ pressure on the upstream side of the membrane from TMAP8 calculation and the experimental data.

!media comparison_val-2e.py
       image_name=val-2e_comparison_mixture_diffusion.png
       style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
       id=val-2e_comparison_mixture_diffusion
       caption= Comparison of the D$_2$, H$_2$, and HD flux versus effective D$_2$ pressure on the upstream side of the membrane from TMAP8 calculation using lawdep condition and the experimental data.

!media comparison_val-2e.py
       image_name=val-2e_comparison_mixture_diffusion_recombination.png
       style=width:50%;margin-bottom:2%;margin-left:auto;margin-right:auto
       id=val-2e_comparison_mixture_diffusion_recombination
       caption= Comparison of the D$_2$, H$_2$, and HD flux versus effective D$_2$ pressure on the upstream side of the membrane from TMAP8 calculation using ratedep condition and the experimental data.

## Input files

!style halign=left

> The input file for this case can be generated by [/val-2ea.i], [/val-2ed.i], and [/val-2ee.i]. The file are also used as test in TMAP8 at [/val-2e/tests].

Reviewer: ```suggestion
For this case, the main input files are [/val-2ea.i], [/val-2ed.i], and [/val-2ee.i]. Sub-cases `ver-1eb` and `ver-1ec` are adapted from [/val-2ea.i], as it done in [/val-2e/tests]. The file are also used as tests in TMAP8 at [/val-2e/tests].
```

| $n$ | exponent for the relation of pressure and concentration | 0.9297         | -                     | [!cite](kizu2001co) |
| $K_{r,D_2}$ | D$_2$ recombination coefficient | 2.502$\times 10^{-24} / \sqrt{4T} \exp(11836/T)$        | m$^4$/s              | [!cite](ambrosek2008verification) |
| $K_{r,H_2}$ | H$_2$ recombination coefficient | 2.502$\times 10^{-24} / \sqrt{2T} \exp(11836/T)$        | m$^4$/s              | [!cite](ambrosek2008verification) |
| $K_{r,HD}$ | HD recombination coefficient | 2.502$\times 10^{-24} / \sqrt{3T} \exp(11836/T)$        | m$^4$/s              | [!cite](ambrosek2008verification) |
| $K_{d,D_2}$ | D$_2$ dissociation coefficient | 2.1897$\times 10^{22} / \sqrt{4T}$     | molecular/m$^2$/Pa              | [!cite](ambrosek2008verification) |
| $K_{d,H_2}$ | H$_2$ dissociation coefficient | 2.1897$\times 10^{22} / \sqrt{2T}$     | molecular/m$^2$/Pa              | [!cite](ambrosek2008verification) |
| $K_{d,HD}$ | HD dissociation coefficient | 2.1897$\times 10^{22} / \sqrt{3T}$     | molecular/m$^2$/Pa              | [!cite](ambrosek2008verification) |

!alert warning title=Typo in [!cite](ambrosek2008verification)

> There are typos on the equations for hydrogen diffusivity, recombination and dissociation coefficients in the input files from TMAP7 in val-2ed and val-2ee. The correct values are provided in [val-2e_parameters] and used in TMAP8. The pre-factor for hydrogen diffusivity is 3.728 $\times 10^{-4}$ m$^2$/s instead of 2.636 $\times 10^{-4}$ m$^2$/s. The activation energy for recombination coefficient is 11836 K instead of -11836 K. The pre-factors for dissociation coefficient are 2.1897 $\times 10^{22} / \sqrt{4T}$, 2.1897 $\times 10^{22} / \sqrt{2T}$, and 2.1897 $\times 10^{22} / \sqrt{3T}$ molecular/m$^2$/Pa instead of 2.1897 $\times 10^{22} / \sqrt{8T}$, 2.1897 $\times 10^{22} / \sqrt{4T}$, and 2.1897 $\times 10^{22} / \sqrt{6T}$ molecular/m$^2$/Pa.

Reviewer: ```suggestion
There are typos on the equations for hydrogen diffusivity, recombination and dissociation coefficients in the input files from TMAP7 in val-2ed and val-2ee. The correct values are provided in [val-2e_parameters] and used in TMAP8. The pre-factor for hydrogen diffusivity used in TMAP8 is 3.728 $\times 10^{-4}$ m$^2$/s instead of 2.636 $\times 10^{-4}$ m$^2$/s as used in the input file published in [!cite](ambrosek2008verification). The activation energy for recombination coefficient used in TMAP8 is 11836 K instead of the -11836 K value shown in Eq. (71) in [!cite](ambrosek2008verification) - note that the input files in [!cite](ambrosek2008verification) use the proper value. The values of the species molecular weight in amu are also updated from the input files shown in [!cite](ambrosek2008verification).
```
